File: LeetCode-Java/First Missing Positive.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        int index = 0;
        while(index < A.length)
        {
            if(A[index] <= 0 || A[index] > A.length)
            {
                A[index] = -1;
                index ++;
            }
            else if(A[index] == index + 1)
            {
                index ++;
            }
            else if(A[index] != index + 1)
            {
                if(A[A[index] - 1] == A[index])
                {
                    A[index] = -1;
                    index ++;
                }
                else
                {
                    int temp = A[index];
                    A[index] = A[A[index] - 1];
                    A[temp - 1] = temp;
                }
            }
        }
        for(index = 0;index < A.length;index ++)
        {
            if(A[index] != index + 1)
                return index + 1;
        }
        return index + 1;
    }
}

File: Leetcode-cpp/first_missing_positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] <= 0 || nums[i] - 1 >= nums.size() || nums[i] - 1 == i) continue;
            int temp = nums[i];
            if (nums[temp - 1] == nums[i]) continue;
            nums[i] = nums[temp - 1];
            nums[temp - 1] = temp;
            i--;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] - 1 != i) return i + 1;
        }
        return nums.size() + 1;
    }
};